File: packages/contracts/src/security-shift.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const securityShiftStatusSchema = z.enum(["ACTIVE", "COMPLETED"]);

export type SecurityShiftStatus = z.infer<typeof securityShiftStatusSchema>;

export const endSecurityShiftInputSchema = z
  .object({
    notes: z.string().trim().max(1000).optional(),
  })
  .strict();

export type EndSecurityShiftInput = z.infer<typeof endSecurityShiftInputSchema>;

export const securityShiftSchema = z.object({
  id: z.string().uuid(),
  officerName: z.string().min(1),
  status: securityShiftStatusSchema,
  startedAt: z.string().datetime({ offset: true }),
  endedAt: z.string().datetime({ offset: true }).nullable(),
  notes: z.string().nullable(),
});

export type SecurityShift = z.infer<typeof securityShiftSchema>;

export const securityShiftResponseSchema = dataEnvelopeSchema(
  z.object({ shift: securityShiftSchema.nullable() }),
);

export const securityShiftMutationResponseSchema = dataEnvelopeSchema(
  z.object({ shift: securityShiftSchema }),
);

export type SecurityShiftResponse = z.infer<typeof securityShiftResponseSchema>;
export type SecurityShiftMutationResponse = z.infer<typeof securityShiftMutationResponseSchema>;

File: apps/mobile/lib/features/security_shift/domain/security_shift.dart
enum SecurityShiftStatus {
  active('ACTIVE'),
  completed('COMPLETED');

  const SecurityShiftStatus(this.apiValue);

  final String apiValue;

  static SecurityShiftStatus fromApi(Object? value) {
    if (value is! String) {
      throw const FormatException('Security shift status is invalid.');
    }
    for (final status in values) {
      if (status.apiValue == value) return status;
    }
    throw FormatException('Unsupported security shift status: $value');
  }
}

/// Mirrors `SecurityShift` in `packages/contracts/src/security-shift.ts`.
class SecurityShift {
  const SecurityShift({
    required this.id,
    required this.officerName,
    required this.status,
    required this.startedAt,
    required this.endedAt,
    required this.notes,
  });

  final String id;
  final String officerName;
  final SecurityShiftStatus status;
  final DateTime startedAt;
  final DateTime? endedAt;
  final String? notes;

  factory SecurityShift.fromJson(Map<String, dynamic> json) {
    final endedAt = json['endedAt'] as String?;
    return SecurityShift(
      id: json['id'] as String,
      officerName: json['officerName'] as String,
      status: SecurityShiftStatus.fromApi(json['status']),
      startedAt: DateTime.parse(json['startedAt'] as String),
      endedAt: endedAt != null ? DateTime.parse(endedAt) : null,
      notes: json['notes'] as String?,
    );
  }
}
